File: Data Structures & Algorithms/subsets/submission-2.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(0, nums, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int start, int[] nums, List<Integer> path, List<List<Integer>> result) {
        // we will create a copy of path and then append it to the result, as path refence can change with every iteration
        result.add(new ArrayList<>(path));

        // we are looping from start not 0 as we don't want to include duplicates
        for (int i = start; i < nums.length; i++) {
            path.add(nums[i]); // choose
            backtrack(i + 1, nums, path, result); // backtrack
            path.remove(path.size() - 1); // undo 
        }
    }
}

File: Data Structures & Algorithms/permutations/submission-2.go
func permute(nums []int) [][]int {

    result := make([][]int, 0)
    path := make([]int, 0)
    used := make([]bool, len(nums))
    
    backtrack(nums, path, used, &result)

    return result;

}

func backtrack(nums []int, path []int, used []bool, result *[][]int) {
    if len(path) == len(nums) {
        temp := make([]int, len(nums))
        copy(temp, path)
        *result = append(*result, temp)
    }

        for i := 0; i < len(nums); i++ {
            if used[i] {
                continue
            }
            path = append(path, nums[i])
            used[i] = true;

            backtrack(nums, path, used, result)
            used[i] = false
            path = path[:len(path) - 1]
        }
}


